File: backend/repositories/settings_repository.py
"""SQLite repository functions for durable dashboard settings."""

from datetime import datetime

from backend.config import DEFAULT_SYMBOL
from backend.database import db_connect
from backend.utils import normalize_symbol


DEFAULT_STOCK_SYMBOL_KEY = "default_stock_symbol"
DEFAULT_STOCK_NAME_KEY = "default_stock_name"


def init_settings_db() -> None:
    """Create the settings table and seed the configured default stock."""
    now = datetime.now().isoformat(timespec="seconds")
    with db_connect() as connection:
        connection.execute(
            """
            CREATE TABLE IF NOT EXISTS app_settings (
                key TEXT PRIMARY KEY,
                value TEXT NOT NULL DEFAULT '',
                updated_at TEXT NOT NULL
            )
            """
        )
        connection.execute(
            """
            INSERT OR IGNORE INTO app_settings (key, value, updated_at)
            VALUES (?, ?, ?)
            """,
            (DEFAULT_STOCK_SYMBOL_KEY, normalize_symbol(DEFAULT_SYMBOL), now),
        )
        connection.execute(
            """
            INSERT OR IGNORE INTO app_settings (key, value, updated_at)
            VALUES (?, ?, ?)
            """,
            (DEFAULT_STOCK_NAME_KEY, "", now),
        )


def get_default_stock_sync() -> dict[str, str]:
    """Read the persisted default stock from settings storage."""
    with db_connect() as connection:
        rows = connection.execute(
            """
            SELECT key, value
            FROM app_settings
            WHERE key IN (?, ?)
            """,
            (DEFAULT_STOCK_SYMBOL_KEY, DEFAULT_STOCK_NAME_KEY),
        ).fetchall()
    values = {row["key"]: row["value"] for row in rows}
    symbol = values.get(DEFAULT_STOCK_SYMBOL_KEY) or DEFAULT_SYMBOL
    return {
        "symbol": normalize_symbol(symbol),
        "name": values.get(DEFAULT_STOCK_NAME_KEY) or "",
    }


def set_default_stock_sync(symbol: str, name: str = "") -> dict[str, str]:
    """Persist a stock as the default dashboard symbol."""
    normalized = normalize_symbol(symbol)
    now = datetime.now().isoformat(timespec="seconds")
    with db_connect() as connection:
        connection.executemany(
            """
            INSERT INTO app_settings (key, value, updated_at)
            VALUES (?, ?, ?)
            ON CONFLICT(key) DO UPDATE SET
                value = excluded.value,
                updated_at = excluded.updated_at
            """,
            [
                (DEFAULT_STOCK_SYMBOL_KEY, normalized, now),
                (DEFAULT_STOCK_NAME_KEY, name or "", now),
            ],
        )
    return {"symbol": normalized, "name": name or ""}

File: backend/trade-watchlist/src/main/java/com/tradeagent/watchlist/repository/SettingsRepository.java
package com.tradeagent.watchlist.repository;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.stereotype.Repository;

import com.tradeagent.market.dto.StockIdentity;
import com.tradeagent.market.util.StockSymbols;
import com.tradeagent.watchlist.config.WatchlistProperties;
import com.tradeagent.watchlist.repository.mapper.SettingsMapper;

/** PostgreSQL repository for durable dashboard settings. */
@Repository
public class SettingsRepository {
    private static final String DEFAULT_STOCK_SYMBOL_KEY = "default_stock_symbol";
    private static final String DEFAULT_STOCK_NAME_KEY = "default_stock_name";

    private final SettingsMapper settingsMapper;
    private final WatchlistProperties properties;

    /** Create the repository with MyBatis settings access and app defaults. */
    public SettingsRepository(SettingsMapper settingsMapper, WatchlistProperties properties) {
        this.settingsMapper = settingsMapper;
        this.properties = properties;
    }

    /** Create settings storage and seed the configured default stock. */
    public void init() {
        settingsMapper.createSettingsTable();
        settingsMapper.ensureTimestampDefaults();
        settingsMapper.insertSettingIfAbsent(DEFAULT_STOCK_SYMBOL_KEY,
                StockSymbols.normalizeSymbol(properties.defaultSymbol()));
        settingsMapper.insertSettingIfAbsent(DEFAULT_STOCK_NAME_KEY, "");
    }

    /** Read the persisted default stock from settings storage. */
    public StockIdentity getDefaultStock() {
        List<Map<String, Object>> rows = settingsMapper.selectSettings(
                List.of(DEFAULT_STOCK_SYMBOL_KEY, DEFAULT_STOCK_NAME_KEY));
        Map<String, String> values = rows.stream()
                .collect(Collectors.toMap(row -> String.valueOf(row.get("key")), row -> String.valueOf(row.get("value"))));
        String symbol = values.getOrDefault(DEFAULT_STOCK_SYMBOL_KEY, properties.defaultSymbol());
        String name = values.getOrDefault(DEFAULT_STOCK_NAME_KEY, "");
        return new StockIdentity(StockSymbols.normalizeSymbol(symbol), name, "");
    }

    /** Persist a stock as the dashboard default and return the saved value. */
    public StockIdentity setDefaultStock(String symbol, String name) {
        String normalized = StockSymbols.normalizeSymbol(symbol);
        String cleanName = name == null ? "" : name;
        settingsMapper.upsertSetting(DEFAULT_STOCK_SYMBOL_KEY, normalized);
        settingsMapper.upsertSetting(DEFAULT_STOCK_NAME_KEY, cleanName);
        return new StockIdentity(normalized, cleanName, "");
    }
}
